### PR TITLE
🎉 Source Shopify: Add privileges validation

### LINKED ISSUE #8235: Source Shopify connection check: Before starting sync, verify input token has permissions to access all resources
## Tell us about the problem you're trying to solve
When creating API tokens, Shopify allows users very granular permissions over which resources the API token has access to: 

Screen Shot 2021-11-24 at 8 46 20 AM

Unfortunately this means that users may accidentally forget to enable permissions for some streams which they selected. This results in a non-obvious error here like this: `requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://<>.myshopify.com/admin/api/2021-07/price_rules.json?limit=250&order=updated_at+asc&updated_at_min=2020-01-01
`

Ideally we can tell the user their API token is missing permissions during the connection check stage. However at that stage we don't know which resources they want to replicate. So it would not make sense to fail the connection check if they don't have access to let's say, `price_rules` unless we know they want to replicate `price_rules`, which we only do once they actually run a sync. 

## Describe the solution you’d like
When a user reads data from Shopify, before syncing data, run a stream-specific connection check which verifies the input API token has permissions to access. If the API token doesn't have access for a particular stream (i.e: if it throws 403) then tell the user in clear terms which streams they need to enable access for.

### PR DESCRIPTION
**Pull request by lazebnyi**
## What
Before starting sync, verify input token has permissions to access all resources - #8235 

## How
Added validation before reading.

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-shopify/source_shopify/utils.py [hunk 1]
```diff
@@ -9,9 +9,21 @@
 
 import requests
 
+SCOPES_MAPPING = {
+    "read_customers": ["Customers"],
+    "read_orders": ["Orders", "AbandonedCheckouts", "Transactions", "Fulfillments", "OrderRefunds", "OrderRisks"],
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-shopify/source_shopify/utils.py:14]: `read_checkouts` should correspond to `AbandonedCheckouts` stream, as of the part of `read_orders`, please correct this.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 2]
```diff
@@ -481,3 +471,43 @@ def streams(self, config: Mapping[str, Any]) -> List[Stream]:
             Fulfillments(config),
             Shop(config),
         ]
+
+    def read(
+        self, logger: AirbyteLogger, config: Mapping[str, Any], catalog: ConfiguredAirbyteCatalog, state: MutableMapping[str, Any] = None
+    ) -> Iterator[AirbyteMessage]:
+        user_scopes = self.get_user_scopes(config)
+
+        always_permitted_streams = ["Metafields", "Shop"]
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-shopify/source_shopify/source.py:480]: Can we rename this to something like `base_streams` ?

**Review by bazarnov — APPROVED** [review]
Looks great, with small comments.

File: airbyte-integrations/connectors/source-shopify/source_shopify/utils.py (hunk 1 above)
Comment (lazebnyi) [reply airbyte-integrations/connectors/source-shopify/source_shopify/utils.py:14]: According to [doc](https://shopify.dev/api/usage/access-scopes) `AbandonedCheckouts` correspond only to `read_orders`.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py (hunk 2 above)
Comment (lazebnyi) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:480]: Metafields and Shop are not base streams. I think the `always_permitted_streams` naming better explain the main idea.

**Review by antixar — CHANGES_REQUESTED** [review]
Can we move this validation to the `streams` function? And as result this connector will provide available streams only

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 3]
```diff
@@ -481,3 +471,43 @@ def streams(self, config: Mapping[str, Any]) -> List[Stream]:
...
+            raise PermissionError(massage)
+
+        yield from super().read(logger, config, catalog, state)
+
+    @staticmethod
+    def get_user_scopes(config):
+        session = requests.Session()
+        headers = ShopifyAuthenticator(config).get_auth_header()
+        response = session.get(f"https://{config['shop']}.myshopify.com/admin/oauth/access_scopes.json", headers=headers).json()
+        return response["access_scopes"]
```
[the remark this answers is not in the document]
Comment (lazebnyi) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:509]: After we shift the main part to the stream method. Yes, we can use the authenticator from config. 
No, we could not use `url_base` without parsing it.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 4]
```diff
@@ -443,9 +433,9 @@ def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) ->
         """
         config["authenticator"] = ShopifyAuthenticator(config)
         try:
-            responce = list(Shop(config).read_records(sync_mode=None))
-            # check for the shop_id is present in the responce
-            shop_id = responce[0].get("id")
+            response = list(Shop(config).read_records(sync_mode=None))
+            # check for the shop_id is present in the response
+            shop_id = response[0].get("id")
             if shop_id is not None:
```
[the remark this answers is not in the document]
Comment (lazebnyi) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:439]: updated.

**Review by antixar — APPROVED** [review]
LGTM
Please publish your changes

**Pull request merged**